Add/rest api/block directory

tellyworth opened the pull request:
Here's a first attempt at moving the block directory API endpoints into core. Still WIP.

Some known issues:
* Schemas are not working (`test_search_schema()` fails).
* The endpoint structure isn't ideal, but I've stuck with the same structure to ensure the Gutenberg plugin can switch between them.
* Unit tests need more coverage, especially for the install and uninstall endpoints.

Plenty of unknown issues too.

tellyworth committed:
Add simple unit tests

tellyworth committed:
Move WP_REST_Block_Directory_Controller to core endpoint folder

Copied from https://github.com/WordPress/gutenberg/pull/17669

tellyworth committed:
Load the endpoint class

tellyworth committed:
Switch to core namespace

tellyworth committed:
Suppress errors and instead return WP_Error

The `plugins_api()` function triggers PHP errors on network failures, in addition to returning error objects. This suppresses the trigger, but still ensures the WP_Error object is returned.

tellyworth committed:
Remove debug output

tellyworth committed:
Fix test assumption

tellyworth committed:
Add test_install_unreachable()

tellyworth committed:
We should be using POST for the install endpoint

tellyworth committed:
Add some more simple install tests

tellyworth committed:
Add test_search_schema() - WIP

Still incomplete. Currently fails, need to add further properties as available.

tellyworth committed:
Add get_item_schema()

Includes a rough draft of the schema.

tellyworth committed:
Cast counter output to ints

See https://github.com/WordPress/gutenberg/pull/17594

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+			return $filesystem_available;
+		}
+
+		$api = @plugins_api(
```
Is the @ required here?

tellyworth replied:
I think so. See test_install_unreachable(): https://github.com/tellyworth/wordpress-develop/pull/1/files#diff-cf39fadfe2b4871f91408969f3b5efaeR87

Without suppressing the error, `plugins_api()` triggers an `E_USER_WARNING` or `E_USER_NOTICE` and the test fails. Happy to handle it differently if there's a better way.

dd32 replied:
I don't think it's proper to include it in the API endpoint then, perhaps in the unit test function which ultimately calls the functions that throw the warning.

Alternatively, see [Example 2.12: Expecting a PHP error using @expectedException](https://phpunit.de/manual/6.5/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.exceptions.examples.ExceptionTest.php) but I don't know if that works with the `_USER_` variant of warnings..

tellyworth replied:
Yeah good point, I've changed it in b2682147f6 to suppress in the tests rather than the API code for both instances. Seems unnecessary to specifically test for the exception given that we only care about the WP_Error response.

TimothyBJacobs replied:
Looks like the warning variant could be caught using `PHPUnit\Framework\Error\Warning` or `PHPUnit_Framework_Error_Warning` as the exception class.

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+		require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		$response = @plugins_api(
```
@ required here?

tellyworth replied:
Fixed in b2682147f6.

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+		}
+
+		// Plugin's description, not description in block.json.
+		$block->description = wp_trim_words( wp_strip_all_tags( $plugin['description'] ), 30, '...' );
```
`wp_trim_words()` internally calls `wp_strip_all_tags()` so that can be removed here.

tellyworth replied:
Fixed this one, thanks.

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+		$block->author = wp_strip_all_tags( $plugin['author'] );
+
+		// Plugin's icons or icon in block.json.
+		$block->icon = isset( $plugin['icons']['1x'] ) ? $plugin['icons']['1x'] : 'block-default';
```
Any reason this is using `1x`? That might not work for SVG icons.

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+		$block->assets = array();
+		foreach ( $plugin['block_assets'] as $asset ) {
+			// TODO: Return from API, not client-set.
+			$block->assets[] = 'https://plugins.svn.wordpress.org/' . $plugin['slug'] . $asset;
```
We should really look at returning that from the API side before this is merged into core

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+
+		$block->humanized_updated = sprintf(
+			/* translators: %s: Human-readable time difference. */
+			__( '%s ago', 'gutenberg' ),
```
textdomain not required for core.

tellyworth replied:
Fixed in 5163f8a4bc.

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+				),
+				'author'              => array(
+					'description'       => __( "The WordPress.org username of the block author." ),
+					'type'              => 'integer',
```
string

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+				),
+				'active_installs'     => array(
+					'description'       => __( "The number sites that have activated this block." ),
+					'type'              => 'integer',
```
string

tellyworth replied:
Fixed in 5163f8a4bc

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+					'context'           => array( 'view' ),
+				),
+				'icon' => array(
+					'description'     => __( "The WordPress.org username of the block author." ),
```
icon != username

tellyworth replied:
Fixed in 5163f8a4bc

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+				'type' => 'string',
+			),
+			'required'          => true,
+			'sanitize_callback' => array( $this, 'sanitize_text_field' ),
```
Might be worth checking to see if `sanitize_callback` is needed here, and/or if that's compatible with `type => array`.

TimothyBJacobs replied:
It seems unnecessary to me. It also seems like it is referencing a method on the controller class, but I'm not seeing the associated method.

tellyworth committed:
Change the prepare function to be more in line with core patterns

Replace parse_block_metadata() with a similar function called prepare_item_for_response(). This uses a data array rather than an object, and returns a WP_REST_Response object.

tellyworth committed:
Remove redundant call

@dd32 pointed out that wp_trim_words() already calls wp_strip_all_tags() internally.

tellyworth committed:
Fix a bug where prepare_item_for_response() wasn't called correctly

Also use prepare_response_for_collection(). This means the returned data is now an array instead of an object, so change the test to suit also.

tellyworth committed:
Fix typo in response key

'assets' not 'asssets'

tellyworth committed:
Suppress PHP errors in the tests, not the API code

Thanks Dion!

tellyworth committed:
Fix textdomains and some Gutenberg references

tellyworth committed:
Fixed some schema errors

Thanks Dion!

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,514 @@
+			'author'                => wp_strip_all_tags( $plugin['author'] ),
+			'icon'                  => ( isset( $plugin['icons']['1x'] ) ? $plugin['icons']['1x'] : 'block-default' ),
+			'assets'                => array(),
+			'humanized_updated'     => sprintf(
```
Just returning the `last_updated` would be closer to how the REST API generally tries to expose data. Ie exposing the underlying value instead of formatted for a specific context. But I think Gutenberg still doesn't have a `human_time_diff` equivalent. So if we need to keep it, I'd just also expose the raw `last_updated` property.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,514 @@
+			$block[ 'assets' ][] = 'https://plugins.svn.wordpress.org/' . $plugin['slug'] . $asset;
+		}
+
+		$response = new WP_REST_Response();
```
We can pass `$block` directly to `WP_REST_Response` as the first parameter. No need for the separate `::set_data()` call.

tellyworth replied:
Fixed in d439787d88

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,514 @@
+	public function get_item_schema() {
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'block',
```
```suggestion
			'title'      => 'block-directory-item',
```

Or something similar. We'll have a blocks endpoint at some point which wouldn't have the same schema.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+ * @package Gutenberg
+ * phpcs:disable
+ */
+class WP_REST_Block_Directory_Controller_Test extends WP_UnitTestCase {
```
```suggestion
class WP_REST_Block_Directory_Controller_Test extends WP_Test_REST_TestCase {
```

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+		global $wp_rest_server;
+		$wp_rest_server = new \WP_REST_Server;
+		$this->server   = $wp_rest_server;
+		do_action( 'rest_api_init' );
```
```suggestion
		do_action( 'rest_api_init', $this->server );
```

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+
+		global $wp_rest_server;
+		$wp_rest_server = new \WP_REST_Server;
+		$this->server   = $wp_rest_server;
```
We need the `protected $server;` definition for the server.

tellyworth replied:
I'm not sure it's needed at all, since I've switched the test functions to use `rest_get_server()->dispatch()`

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+		do_action( 'rest_api_init' );
+
+		$this->controller = new WP_REST_Block_Directory_Controller();
+		$this->controller->register_routes();
```
This shouldn't be necessary, adding the registration in `create_initial_rest_routes` should be sufficient for the tests.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+	function test_should_throw_no_slug_error() {
+		$request = new WP_REST_Request( 'POST', '/wp/v2/block-directory/install', [] );
+
+		$result = $this->controller->install_block( $request );
```
Instead of calling the controller directly, we typically call `rest_do_request()` That way all the server processing occurs like parameter validation.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+
+		$this->prevent_requests_to_host( 'api.wordpress.org' );
+
+		$result = @ $this->controller->get_items( $request );
```
Why do we need the error suppression here?

tellyworth replied:
`get_items()` calls `plugins_api()`, which triggers an `E_USER_WARNING` or `E_USER_NOTICE` in addition to returning a WP_Error object.

Suppressing it in the test seems like a better approach than in the API code. See discussion here: https://github.com/tellyworth/wordpress-develop/pull/1#discussion_r420562537

tellyworth replied:
Fixed in 772d23955f.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+
+		$this->assertEquals( $data['code'], 'rest_user_cannot_view' );
```
This will be able to use `assertErrorResponse` once we change the base class.

tellyworth replied:
Switched to assertErrorResponse in 511bbdb054.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php:
```diff
@@ -0,0 +1,194 @@
+		$data     = $response->get_data();
+
+		// Is this an appropriate status?
+		$this->assertEquals( 500, $response->status );
```
The REST API returns 200 responses for collections, even if no items are found.

tellyworth replied:
This was outdated by other changes, but I've added a test in 1e813a59f3 to take that into account.

TimothyBJacobs commented:
I think `test_search_schema` isn't working because the route isn't being created in `create_initial_rest_routes()`.

tellyworth committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php

Co-authored-by: Timothy Jacobs <timothy@ironbounddesigns.com>

tellyworth committed:
Update tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php

Co-authored-by: Timothy Jacobs <timothy@ironbounddesigns.com>

tellyworth committed:
Update tests/phpunit/tests/rest-api/class-wp-rest-block-directory-controller-test.php

Co-authored-by: Timothy Jacobs <timothy@ironbounddesigns.com>

tellyworth committed:
Register block directory routes in create_initial_rest_routes()

tellyworth committed:
Clean up tests

Use more standard methods including `assertErrorResponse()`

tellyworth committed:
Make sure error responses include a status value

Status codes may need some revision.

tellyworth commented:
> I think `test_search_schema` isn't working because the route isn't being created in `create_initial_rest_routes()`.

It's certainly changed since then, but it's now getting a 404 response instead. Probably something else I've missed in init, will dig further.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php:
```diff
@@ -0,0 +1,518 @@
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		$schema = array(
```
We should also cache this in `$this->schema`. See `WP_REST_Posts_Controller::get_item_schema()`.

tellyworth replied:
Added in 1e813a59f3.

tellyworth committed:
No need for a separate set_data() call here

Thanks Timothy!

tellyworth committed:
Use rest_do_request()

tellyworth committed:
Add a test for empty search results.

tellyworth committed:
Cache the schema in object attribute.

tellyworth committed:
Use expectException() instead of suppressing errors.

Thanks Timothy!

tellyworth committed:
Remove sanitize_callback

It was incorrect, and is probably unneeded anyway.
